File: calumny.py
#!/usr/bin/python
# -*- coding: utf-8 -*-
import cherrypy # requires python-cherrypy3 
import numpy
import random
import time
class rl(object):
    def __init__(self,loc):
        self.loc=numpy.array(loc)
        self.m=numpy.transpose(numpy.array([map(ord,i) for i in open("./dun").read().split("\n")], dtype=numpy.int))                  # mapmaking via mx artist-mode
        self.m[tuple(self.loc)]=ord("@")
        self.v=numpy.array([25,10])      # visible field
        self.s=numpy.array(self.m.shape) # array shape
        self.z=numpy.array([0,0])
    directions={37:numpy.array([-1,0]),38:numpy.array([0,-1]),39:numpy.array([1,0]),40:numpy.array([0,1])}
    def move(self,event):
        if event["keyCode"]!="undefined" and int(event["keyCode"]) in self.directions:
           motion=self.directions[int(event["keyCode"])]
           self.m[tuple(self.loc)]=ord(" ")
           self.loc+=motion
           self.m[tuple(self.loc)]=ord("@")
        return self.rlview()
    def char(self):
        return r'''<div style="position:absolute;left:%dem;top:%dem;"><pre>Name:%s
Location:%d,%d<pre></div>'''%(self.v[0]+2,0,str(self),self.loc[0],self.loc[1])
    def rlview(self):
        a=numpy.minimum(self.s-2*self.v,numpy.maximum(self.z,self.loc-self.v))
        b=a+2*self.v
        return (u"<pre>"+(u"\n".join("".join(map(unichr,self.m[a[0]:b[0],index])) for index in range(a[1],b[1])))+u"</pre>").encode("utf-8")+self.char()
class view(object):
    def __init__(self):
        self.i=0
        self.rls={}
    @cherrypy.expose
    def upload(self,evt):
        self.i+=1
        rli=self.getrli()
        event=dict(par.split(":",1) for par in evt.strip().split("\n"))
        return self.rls[rli].move(event)+("</pre><br><b>%d events ok</b>"%self.i).encode("utf-8")
        return "<b>OK %d</b>"%self.i
    def getrli(self):
        rli=cherrypy.request.cookie.get('rougelikeinstance',None)
        if rli==None:
            cookie = cherrypy.response.cookie
            rli=str(time.time())
            cookie['rougelikeinstance']=rli
            cookie['rougelikeinstance']['path'] = '/'
            cookie['rougelikeinstance']['max-age'] = 3600
            cookie['rougelikeinstance']['version'] = 1
        else:
            rli=rli.value
        if self.rls.get(rli,None) == None:            
            print "RESETTING CHARACTER"
            self.rls[rli]=rl((10,10))
        return rli
    @cherrypy.expose
    def index(self,*args):
        return r'''<html><head>
<style>
pre { font-family: Courier, monospace; border: .1em dotted black; display: table; }
</style>
<script type="text/javascript">
var list=[];
var timeout;
var xhr = new XMLHttpRequest();
var keys=["timeStamp","altKey","shiftKey","ctrlKey","metaKey","keyLocation","type","keyIdentifier","keyCode","charCode","button","clientX","clientY"];
function replac() {
  if (xhr.readyState == 4){
    if(list.length>0) {
        doKey(list.pop());
      }
      document.getElementById('upload').innerHTML=xhr.responseText;
  }
}
function doKey(e) {
  if (xhr.readyState == 4 || xhr.readyState == 0){    
     var params="evt=eventstring:"+e.toString()+"\n";
     for (attr in keys)
        params+=keys[attr]+":"+e[keys[attr]]+"\n";
     xhr.open("POST", "upload",true);
     xhr.setRequestHeader("Content-type", "application/x-www-form-urlencoded");
     xhr.send(params);
     window.clearTimeout(timeout);
     timeout=window.setTimeout ("doKey('TIME')", 1000);
  } else {
     list.push(e)
  }
};
function stop() {
    window.clearTimeout(timeout);
    xhr.onreadystatechange=stop;
}
onload=function()
{
  if(navigator.userAgent.indexOf("Gecko/")==-1)
  { document.onkeydown=doKey; }
  else
  { document.onkeypress=doKey; }
  xhr.onreadystatechange=replac;
  doKey("INIT");
  timeout=window.setTimeout ("doKey('TIME')", 1000);

}
</script>
</head><body><div id="upload"></div><button type="button" onclick="stop()" >stop client</button></body></html>'''
cherrypy.quickstart(view())

File: calumny.rb
require 'java'
Dir.glob("./lib/*.jar").each{|x| require x;}
require 'rlmock' #where the voldemort interface lives
server = org.mortbay.jetty.Server.new(8080)
thread_pool = org.mortbay.thread.QueuedThreadPool.new
thread_pool.min_threads  = 10 
thread_pool.max_threads  = 20
server.set_thread_pool(thread_pool)
context_static = org.mortbay.jetty.servlet.Context.new(server, '/static', 0)
context_static.handler=org.mortbay.jetty.handler.ResourceHandler.new
context_static.resource_base='./static'
context_static.context_path='/static'
class ViewServlet < javax.servlet.http.HttpServlet
  def initialize(rl)
    @gson=com.google.gson.Gson.new() #will use vold.rb eventually. It's faster than gson 
    @rl=rl
    super()
    @response=0
    puts "Ready to serve"
  end
  def doGet(request, response)
    input = request.getParameter('input') #main page
    response.writer.println('<html><head><title>Calumny</title>
<style>div,pre { font-family: Courier, monospace; border: .1em dotted black; display: table; }</style>
<script src="static/dojo.js" type="text/javascript"></script>
<script type="text/javascript">
var img=[]; for(var i in [0,1,2,3,4,5,6,7,8]) {img[i]=new Image();img[i].src="static/tiles/"+i+".png"}
asc=[" ","#",".","w","d","o","@","M","$","t"]; //cache basic replacement images and characters.
var list=[]; var data=null;
var timeout; var timeagain=true; keyconn=null;
xhr=dojo._xhrObj();
function connectkey() { if(keyconn==null) {keyconn=dojo.connect(document.documentElement, "onkeypress", this, doKey);} }
function stop()  { timeagain=false; timer(); dojo.disconnect(keyconn); keyconn=null; list=[]; }
function start()  { timeagain=true; timer(); connectkey();}
function timer()  { window.clearTimeout(timeout); if (timeagain) {timeout=window.setTimeout ("doEvt([\'time\',1000])", 1000);} }
function logout() { doEvt(["logout"]); }
function login() { doEvt(["login",dojo.byId("lu").value,dojo.byId("lp").value]); }
function creat() { doEvt(["create",dojo.byId("cu").value,dojo.byId("cp").value]); }
function asciiview() {
   view=data["view"]
   if(view==null)
   { dojo.byId("upload").innerHTML="<div id=error>Error: No View Data</div>" }
   else {  
     for(var i=0; i<view.length;i++)
       {for(var j=0; j<view[i].length;j++)
         { var contents=view[i][j];
           view[i][j]=asc[contents[contents.length-1]];}
           view[i]=view[i].join(""); }
     dojo.byId("upload").innerHTML=("<pre>"+view.join("\n")+"<pre>"+data["after"]); }
}
function setdata() { dojo.byId("upload").innerHTML=data; }
function canvasview() {}
function webglview() {}
function replac() {
  if (xhr.readyState == 4){
      ret=JSON.parse(xhr.responseText); // dojo fromjson tojson replaced due to speed
      //dojo.byId("upload").innerHTML=ret["ih"];
      data=ret["data"];
      eval(ret["js"]);
      if(list.length>0) { doEvt(list.pop()); }
  }
}
function doEvt(e) {
  if (xhr.readyState == 4 || xhr.readyState == 0){
     xhr.open("POST","evt/"+e[0],true);
     xhr.setRequestHeader("Content-type", "application/x-www-form-urlencoded");
     xhr.send("evt="+JSON.stringify(e));     
     timer();
  } else { list.push(e); }
};
function doKey(e) { doEvt(["key",e["charOrCode"]]); }
dojo.addOnLoad(function()
{ connectkey();
  xhr.onreadystatechange=replac; 
  doEvt(["init"]); });
</script></head>
<body><h5>calumny</h5><div id="upload"></div><div id="control">
<button type="button" onclick="stop()" >stop client</button>
<button type="button" onclick="start()" >start client</button>
<button type="button" onclick="logout()" >logout client</button>
</div></body></html>')
    request.handled = true
  end
  def doPost(request, response)
    start = Time.now
    begin
      input=fromJson(request.getParameter('evt')).to_a # p input
      cookies = request.cookies.to_a.select{|x| x.name=="rougelikeclient"}
      if cookies.length>0 and (username=@rl.username((id=cookies[0].value)))!=nil #if is logged in and exists
        if(input[0]=="key" and (motion=input[1])!=nil)
          @rl.move(input[1],id)
        end
        re=@rl.rlview(id)
        response.writer.println(@gson.toJson({:js=>'start();asciiview();',:data=>{:view=>re,:after=>"Response Time=#{(Time.now - start).to_s}\nUsername=#{username}\nLocation=#{@rl.location(id).to_s}"}}))
      elsif ( input[0]=="login" or input[0]=="create" ) and input.length==3 and (username=input[1].gsub(/\W+/,""))!="" and (password=input[2])!="" #if user does not exist or is not logged in, but we got a login string
        puts "Logging in"
        val=@rl.spawn([username,password]) #login or create user, depending if user was known
        cookie=javax.servlet.http.Cookie.new("rougelikeclient",val)
        cookie.path="/"
        cookie.max_age=3600
        cookie.version=1
        response.addCookie(cookie)
        re=@rl.rlview(val) 
        response.writer.println(@gson.toJson({:js=>'start();asciiview();',:data=>{:view=>re,:after=>"Response Time=#{(Time.now - start).to_s}\nUsername=#{username}\nLocation=#{@rl.location(val).to_s}"}}))
      else # show login
        response.writer.println(@gson.toJson({:data=>'
<div id="up">
  <h2>Log In</h2>
  <div>Username:<input type="text" id="lu"></div>
  <div>Password: <input type="password" name="pwd" id="lp"></div>
<div id="l"><button type="button" onclick="login()">Log in!</button></div></div>
  <div id="create">
  <h2>Create Account</h2>
  <div>Username:<input type="text" id="cu"></div>
  <div>Password: <input type="password" name="pwd" id="cp"></div>
<div id="l"><button type="button" onclick="creat()">Create Account!</button></div></div>
', :js=>'stop();setdata();'}))
      end
      request.handled = true
    rescue Exception=>e
      puts "Exception: #{e}\nBacktrace:#{e.backtrace}"
    end    
  end
end
puts "Launching server"
context=org.mortbay.jetty.servlet.Context.new(server, '/', org.mortbay.jetty.servlet.Context::NO_SESSIONS)
holder =org.mortbay.jetty.servlet.ServletHolder.new(ViewServlet.new(RLMockup.new))
context.addServlet(holder, '/')
server.start()
